File: py/Brain.py
# Python
#

import serial

MAX_PACKET_LENGTH = 32
EEG_POWER_BANDS = 8

class Brain:
    def __init__(self, brain_stream):

        self.brain_stream = brain_stream
        self.fresh_packet = False
        self.in_packet = False
        self.packet_index = 0
        self.packet_length = 0
        self.eeg_power_length = 0
        self.has_power = False
        self.checksum = 0
        self.checksum_accumulator = 0

        self.signal_quality = 200
        self.attention = 0
        self.meditation = 0

        self.last_byte = None
        self.latest_error = ""
        self.csv_buffer = ""
        self.packet_data = [0] * MAX_PACKET_LENGTH
        self.eeg_power = [0] * EEG_POWER_BANDS

        self.clear_eeg_power()

    def update(self):
        if self.brain_stream.in_waiting:
            latest_byte = ord(self.brain_stream.read(1))  # read a byte of data

            # Build a packet if we know we're and not just listening for sync bytes.
            if self.in_packet:
                # First byte after the sync bytes is the length of the upcoming packet.
                # 获取数据包长度
                if self.packet_index == 0:
                    self.packet_length = latest_byte
                    # Catch error if packet is too long
                    if self.packet_length > MAX_PACKET_LENGTH:
                        # Packet exceeded max length
                        # Send an error
                        self.latest_error = f"Error: Packet length {self.packet_length} is too long."
                        self.in_packet = False

                elif self.packet_index <= self.packet_length:
                    # Run of the mill data bytes.

                    # Print them here

                    # Store the byte in an array for parsing later.
                    # 读取到字节存入数据包数组
                    self.packet_data[self.packet_index - 1] = latest_byte

                    # Keep building the checksum.
                    self.checksum_accumulator += latest_byte

                elif self.packet_index > self.packet_length:
                    # 数据包末尾是checksum byte.
                    # We're at the end of the data payload.

                    # Check the checksum.
                    self.checksum = latest_byte
                    self.checksum_accumulator = 255 - self.checksum_accumulator

                    # Do they match?
                    if self.checksum == self.checksum_accumulator:
                        if self.parse_packet():  # Parse the packet
                            self.fresh_packet = True
                        else:
                            # Parsing failed, send an error.
                            self.latest_error = "Error: Could not parse packet."
                            # good place to print the packet if debugging
                    else:
                        # Checksum mismatch, send an error.
                        self.latest_error = "Error: Checksum mismatch."
                        # good place to print the packet if debugging

                    # End of packet
                    # Reset, prep for next packet
                    self.in_packet = False

                self.packet_index += 1

            # 1. looking for the start of the packet
            if latest_byte == 170 and self.last_byte == 170 and not self.in_packet:
                # 170 (0xAA) 是数据包的同步字节。检测到连续的两个 170 字节表示数据包的开始，进入数据包读取状态
                # Start of packet
                self.in_packet = True
                self.packet_index = 0
                self.checksum_accumulator = 0

            # keep track of the last byte so we can find the sync byte pairs.
            self.last_byte = latest_byte

        if self.fresh_packet:
            self.fresh_packet = False
            return True
        else:
            return False

    def clear_packet(self):
        self.packet_data = [0] * MAX_PACKET_LENGTH

    def clear_eeg_power(self):
        # Zero the power bands.
        self.eeg_power = [0] * EEG_POWER_BANDS

    def parse_packet(self):
        # Parse the packet
        # Loop through the packet, extracting data.
        # Based on mindset_communications_protocol.pdf from the Neurosky Mindset SDK.
        # Returns true if passing succeeds
        self.has_power = False
        parse_success = True

        # clear the eeg power to make sure we're honest about missing values
        self.clear_eeg_power()

        i=0
        # iterate over each byte in the data packet
        while i < self.packet_length:
            if self.packet_data[i] == 0x02:  # signal quality
                self.signal_quality = self.packet_data[i + 1]
                i += 1
            elif self.packet_data[i] == 0x04:  # attention
                self.attention = self.packet_data[i + 1]
                i += 1
            elif self.packet_data[i] == 0x05:  # meditation
                self.meditation = self.packet_data[i + 1]
                i += 1
            elif self.packet_data[i] == 0x83:  # ASIC_EEG_POWER
                # ASIC_EEG_POWER: eight big-endian 3-uint8_t unsigned integer values representing
                # delta, theta, low-alpha high-alpha, low-beta, high-beta, low-gamma, and mid-gamma EEG band power values

                # The next uint8_t sets the length, usually 24 (Eight 24-bit numbers... big endian?)
                # We do not use this value so let's skip it and just increment i
                i += 1
                for j in range(EEG_POWER_BANDS):
                    self.eeg_power[j] = (self.packet_data[i + 1] << 16) | (self.packet_data[i + 2] << 8) | self.packet_data[i + 3]
                    i += 3
                self.has_power = True
                # This seems to happen once during start-up on the force trainer. Strange. Wise to wait a couple of packets before
                # you start reading.
            elif self.packet_data[i] == 0x80:
                # We dont' use this value so let's skip it and just increment i
                # 跳过原始脑波数据
                i += 2
            else:
                # Unknown code
                parse_success = False
                break
            i += 1

        return parse_success

    def read_errors(self):
        return self.latest_error

    # Keeping this around for debug use
    def read_csv(self):
        if self.has_power:
            self.csv_buffer = f"{self.signal_quality},{self.attention},{self.meditation}," + ",".join(map(str, self.eeg_power))
        else:
            self.csv_buffer = f"{self.signal_quality},{self.attention},{self.meditation}"
        return self.csv_buffer

        # if self.has_power:
        #     self.csv_buffer = ",".join(map(str, self.eeg_power))
        # return self.csv_buffer

    def read_signal_quality(self):
        return self.signal_quality

    def read_attention(self):
        return self.attention

    def read_meditation(self):
        return self.meditation

    def read_eeg_power(self):
        return self.eeg_power

    def read_delta(self):
        return self.eeg_power[0]

    def read_theta(self):
        return self.eeg_power[1]

    def read_low_alpha(self):
        return self.eeg_power[2]

    def read_high_alpha(self):
        return self.eeg_power[3]

    def read_low_beta(self):
        return self.eeg_power[4]

    def read_high_beta(self):
        return self.eeg_power[5]

    def read_low_gamma(self):
        return self.eeg_power[6]

    def read_mid_gamma(self):
        return self.eeg_power[7]


if __name__ == '__main__':
    port_num = 9600
    brain_stream = serial.Serial('COM3', port_num)  # 串口端口
    brain = Brain(brain_stream)

    while True:
        if brain.update():
            print("New data packet received:")
            print(brain.read_csv())
        else:
            print("No new data packet.")

File: C#/Brain.cs
using System;
using System.IO.Ports;

public class Brain{

    private const int MAX_PACKET_LENGTH = 32; // 定义最大数据包长度
    private const int EEG_POWER_BANDS = 8;    // 定义EEG功率频带数量

    private SerialPort brainStream;
    private byte[] packetData;
    private bool inPacket;
    private byte latestByte;
    private byte lastByte;
    private int packetIndex;
    private int packetLength;
    private byte checksum;
    private byte checksumAccumulator;
    private int eegPowerLength;
    private bool hasPower;
    private string latestError;
    private string csvBuffer;
    private byte signalQuality;
    private byte attention;
    private byte meditation;
    private bool freshPacket;
    private uint[] eegPower;

    public Brain(SerialPort _brainStream){
        brainStream = _brainStream;
        packetData = new byte[MAX_PACKET_LENGTH];
        eegPower = new uint[EEG_POWER_BANDS];
        init();
    }

    private void init(){
        freshPacket = false;
        inPacket = false;
        packetIndex = 0;
        packetLength = 0;
        eegPowerLength = 0;
        hasPower = false;
        checksum = 0;
        checksumAccumulator = 0;

        signalQuality = 200;
        attention = 0;
        meditation = 0;
        ClearEegPower();
    }

    public bool update(){
        // get the number of bytes that can be read in the input buffer
        if (brainStream.BytesToRead > 0) {

            // read one byte of data, returns an integer representing the bytes read.
            latestByte = (byte)brainStream.ReadByte(); 

            if (inPacket) { // Build a packet if we know we're and not just listening for sync bytes.
                if (packetIndex == 0) {
                    packetLength = latestByte;
                    if (packetLength > MAX_PACKET_LENGTH) {
                        latestError = $"ERROR: Packet too long {packetLength}";
                        inPacket = false;
                    }
                } else if (packetIndex <= packetLength) {
                    packetData[packetIndex - 1] = latestByte;
                    checksumAccumulator += latestByte;
                } else if (packetIndex > packetLength) {
                    // We're at the end of the data payload.

                    // Check the checksum.
                    checksum = latestByte;
                    checksumAccumulator = (byte)(255 - checksumAccumulator);

                    // Do they match?
                    if (checksum == checksumAccumulator) {
                        if (ParsePacket()) {
                            freshPacket = true;
                        }
                        else {
                            latestError = "ERROR: Could not parse";
                        }
                    }
                    else {
                        latestError = "ERROR: Checksum";
                    }

                    inPacket = false;
                }

                packetIndex++;
            }

            // Look for the start of the packet 0xAA
            if (latestByte == 170 && lastByte == 170 && !inPacket) {
                inPacket = true;
                packetIndex = 0;
                checksumAccumulator = 0;
            }

            lastByte = latestByte;
        }

        if (freshPacket) {
            freshPacket = false;
            return true;
        }
        else {
            return false;
        }
    }

    private void clearPacket() {
        // private byte[] packetData;
        Array.Clear(packetData, 0, packetData.Length);
    }

    private void clearEegPower() {
        // Zero the power bands.
        Array.Clear(eegPower, 0, eegPower.Length);
    }

    private bool parsePacket() {
        // Loop through the packet, extracting data.
        // Based on mindset_communications_protocol.pdf from the Neurosky Mindset SDK.
        // Returns true if passing succeeds
        hasPower = false;
        bool parseSuccess = true;
        int rawValue = 0;
        ClearEegPower();

        for (int i = 0; i < packetLength; i++) {
            switch (packetData[i]) {
                case 0x02:
                    signalQuality = packetData[++i];
                    break;
                case 0x04:
                    attention = packetData[++i];
                    break;
                case 0x05:
                    meditation = packetData[++i];
                    break;
                case 0x83:  // EEG功率数据标识符
                    i++;
                    for (int j = 0; j < EEG_POWER_BANDS; j++) {
                        eegPower[j] = (uint)((packetData[++i] << 16) | (packetData[++i] << 8) | packetData[++i]);
                    }
                    hasPower = true;
                    break;
                case 0x80:  // 原始脑波数据标识符（未使用）
                    i++;  // 跳过长度字节
                    rawValue = (packetData[++i] << 8) | packetData[++i];
                    break;
                default: // 如果遇到未识别的数据标识符
                    parseSuccess = false;
                    break;
            }
        }
        return parseSuccess;
    }

    public void PrintCSV(){
        // 打印CSV格式的脑电波数据
        string csvOutput = $"{signalQuality},{attention},{meditation}";

        if (hasPower) {
            for (int i = 0; i < eegPower.Length; i++) {
                csvOutput += $",{eegPower[i]}";
            }
        }

        Console.WriteLine(csvOutput);  // 打印输出到控制台或者其他串行流
    }

    public string readErrors(){ 
        return latestError;
    }    
    public string readCSV() {
        if (hasPower) {
            csvBuffer = $"{signalQuality},{attention},{meditation},{string.Join(",", eegPower)}";
        }
        else {
            csvBuffer = $"{signalQuality},{attention},{meditation}";
        }
        return csvBuffer;
    }

    private void printDebug() {
        Console.WriteLine("--- Start Packet ---");
        Console.WriteLine($"Signal Quality: {signalQuality}");
        Console.WriteLine($"Attention: {attention}");
        Console.WriteLine($"Meditation: {meditation}");

        if (hasPower) {
            Console.WriteLine("EEG POWER:");
            Console.WriteLine($"Delta: {eegPower[0]}");
            Console.WriteLine($"Theta: {eegPower[1]}");
            Console.WriteLine($"Low Alpha: {eegPower[2]}");
            Console.WriteLine($"High Alpha: {eegPower[3]}");
            Console.WriteLine($"Low Beta: {eegPower[4]}");
            Console.WriteLine($"High Beta: {eegPower[5]}");
            Console.WriteLine($"Low Gamma: {eegPower[6]}");
            Console.WriteLine($"Mid Gamma: {eegPower[7]}");
        }

        Console.WriteLine($"Checksum Calculated: {checksumAccumulator}");
        Console.WriteLine($"Checksum Expected: {checksum}");
        Console.WriteLine("--- End Packet ---");
    }

    
    public byte readSignalQuality() => signalQuality;
    public byte readAttention() => attention;
    public byte readMeditation() => meditation;
    public uint[] readPowerArray() => eegPower;

    public uint readDelta() => eegPower[0];

    public uint readTheta() => eegPower[1];

    public uint readLowAlpha() => eegPower[2];

    public uint readHighAlpha() => eegPower[3];

    public uint readLowBeta() => eegPower[4];

    public uint readHighBeta() => eegPower[5];

    public uint readLowGamma() => eegPower[6];

    public uint readMidGamma() => eegPower[7];
}

// 主函数类，用于测试 Brain 类
class Program
{
    static void Main()
    {
        // 模拟串行端口通信（例如，COM3）
        string portName = "COM3"; // 替换为你的实际串口端口
        int baudRate = 9600;
        SerialPort brainStream = new SerialPort(portName, baudRate);

        // 打开串口
        brainStream.Open();

        // 创建 Brain 对象
        Brain brain = new Brain(brainStream);

        // 循环读取数据并测试解析功能
        while (true)
        {
            try
            {
                // 更新数据
                if (brain.Update())
                {
                    // 打印 CSV 格式的数据
                    brain.PrintCSV();
                }

                // 为避免死循环
                Thread.Sleep(100); 
            }
            catch (Exception ex)
            {
                Console.WriteLine($"发生错误: {ex.Message}");
                break; // 如果发生错误，退出循环
            }
        }

        // 关闭串口
        brainStream.Close();
    }
}

